File: html5/src/rendering/GraphicsSettings.js
// -----------------------------------------------------------------------------
// GraphicsSettings — every quality knob the engine has, in one place.
//
// The mirror of SexyBiscuit.Engine/Rendering/GraphicsSettings.cs.
//
// One object rather than a scattering of properties on the renderer, the config and
// the canvas, because a settings menu has to be able to read them all, write them all
// and put them in a file. `apply` is the only place that knows where each one lives.
//
// The native-only and web-only groups both exist on both engines on purpose: a web
// export is produced on a desktop, and a preset that lost its maxPixelRatio on the way
// through would ship a phone build at three times the resolution it can fill. The
// capability probe, not this file, decides what a player is shown.
// -----------------------------------------------------------------------------

import presetTable from './graphics-presets.json' with { type: 'json' };
import { setTessellation } from './PrimitiveMesh.js';
import { Camera3D } from './Camera3D.js';

/** How much a shadow costs. The C# enum knows the same list, in order. */
export const ShadowQuality = Object.freeze({
    Off: 'Off', Low: 'Low', Medium: 'Medium', High: 'High', Ultra: 'Ultra',
});

/** How a texture is sampled. Anisotropic falls back where it is unsupported. */
export const TextureFiltering = Object.freeze({
    Point: 'Point', Bilinear: 'Bilinear', Trilinear: 'Trilinear', Anisotropic: 'Anisotropic',
});

/** How much the 2D light map costs, which is mostly its resolution. */
export const Lighting2DQuality = Object.freeze({
    Off: 'Off', Low: 'Low', Medium: 'Medium', High: 'High',
});

/** The key the whole object is stored under, as one JSON blob. */
export const PREFS_KEY = 'Graphics';

export class GraphicsSettings {
    constructor() {
        // --- shared ---
        this.preset = 'medium';
        this.renderScale = 1;
        this.vsync = true;
        this.frameCap = 0;
        this.shadows = ShadowQuality.Low;
        this.shadowDistance = 50;
        this.maxLightsPerObject = 4;
        this.tessellationRadial = 24;
        this.tessellationRings = 16;
        this.drawDistance = 1000;
        this.frustumCulling = true;
        this.sortOpaqueFrontToBack = true;
        this.skybox = true;
        this.particleDensity = 1;
        this.postProcessing = true;
        this.bloom = true;
        this.vignette = false;
        this.colourGrade = true;
        this.scanline = false;
        this.lighting2D = Lighting2DQuality.Medium;
        this.textureFiltering = TextureFiltering.Trilinear;
        this.anisotropy = 4;
        this.streamingRadius = 2;
        this.maxLoadsPerFrame = 1;

        // --- native only ---
        this.msaa = 2;
        this.shadowMapSize = 1024;
        this.fullscreen = false;
        this.resolutionWidth = 0;
        this.resolutionHeight = 0;

        // --- web only ---
        this.maxPixelRatio = 2;
        this.highDpi = true;
        this.antialias = true;
        this.powerPreference = 'high-performance';
        this.shaderPrecision = 'highp';
        this.throttleWhenHidden = true;
    }

    // -------------------------------------------------------------------------
    // Presets
    // -------------------------------------------------------------------------

    /** The preset table, shared with the C# engine as one JSON file. */
    static get presets() { return presetTable.presets; }

    /** The preset a game starts on when nothing else has decided. */
    static get defaultPreset() { return presetTable.default; }

    /** Frame rate to preset, best first. The first threshold cleared wins. */
    static get thresholds() { return presetTable.thresholds; }

    /** A preset by id, case-insensitively, or null. */
    static findPreset(id) {
        const wanted = String(id ?? '').toLowerCase();
        return presetTable.presets.find((p) => p.id.toLowerCase() === wanted) ?? null;
    }

    /** The best preset a machine hitting this frame rate should be offered. */
    static suggest(averageFps) {
        for (const threshold of presetTable.thresholds) {
            if (averageFps >= threshold.minFps) return threshold.preset;
        }
        return presetTable.presets[0]?.id ?? presetTable.default;
    }

    /**
     * How far above the low's own answer an average is allowed to reach.
     *
     * Two rungs is the allowance for a benchmark's own noise -- a single scheduler
     * hiccup should not cost a good machine two quality levels -- while still stopping
     * a badly stuttering one being offered what its average alone would earn.
     */
    static get stutterAllowance() { return 2; }

    /**
     * The preset a machine should be offered, judged on its average and its stutter.
     *
     * Stutter is what a player actually notices. A machine averaging a hundred frames a
     * second with a one-percent low of six is not a hundred-frame machine; it is one
     * that hitches for a whole second in every six, and offering it the preset its
     * average earns would be offering it the preset it just failed at.
     */
    static suggestFor(averageFps, onePercentLowFps) {
        const indexOf = (id) => Math.max(0, presetTable.presets.findIndex((p) => p.id === id));

        const byAverage = indexOf(GraphicsSettings.suggest(averageFps));
        const byLow = indexOf(GraphicsSettings.suggest(onePercentLowFps));

        const last = Math.max(0, presetTable.presets.length - 1);
        const chosen = Math.min(last, Math.max(0, Math.min(byAverage, byLow + GraphicsSettings.stutterAllowance)));

        return presetTable.presets[chosen]?.id ?? presetTable.default;
    }

    /** Overwrites every field from a preset. Unknown ids are ignored. */
    applyPreset(id) {
        const preset = GraphicsSettings.findPreset(id);
        if (!preset) return false;

        Object.assign(this, preset.shared, preset.native, preset.web);
        this.preset = preset.id;
        return true;
    }

    /** A copy, for previewing a change without committing to it. */
    clone() { return Object.assign(new GraphicsSettings(), this); }

    // -------------------------------------------------------------------------
    // Applying
    // -------------------------------------------------------------------------

    /**
     * Pushes every setting to whatever actually owns it.
     *
     * The one place that knows where each knob lives, which is the reason this class
     * exists: they are otherwise spread across the renderer, the mesh cache, the
     * camera, every emitter and the host's own canvas sizing.
     */
    apply(host) {
        applyGlobals(this);
        if (!host) return;

        const renderer = host.renderer3D;
        if (renderer) {
            renderer.enableFrustumCulling = this.frustumCulling;
            renderer.sortOpaqueFrontToBack = this.sortOpaqueFrontToBack;
            renderer.renderSkybox = this.skybox;
            renderer.maxLightsPerObject = Math.min(Math.max(1, this.maxLightsPerObject), 8);
        }

        // Resolution scale on the web is the drawing-buffer size, so it goes through
        // the same pixel-ratio ceiling that stops a phone filling a 3x buffer.
        if (host.config) {
            host.config.highDpi = this.highDpi;
            host.config.maxPixelRatio = this.maxPixelRatio * this.renderScale;
        }
        host.resize?.();
    }
}

/** The settings that live on module state, and so apply with no host at all. */
export function applyGlobals(settings) {
    // setTessellation clears the primitive cache itself, which it has to: the cache is
    // keyed on the shape and not on the segment counts, so meshes built at the old
    // numbers would otherwise survive and the change would be invisible.
    setTessellation(settings.tessellationRadial, settings.tessellationRings);

    const camera = Camera3D.main;
    if (camera) camera.farClip = settings.drawDistance;
}

// -----------------------------------------------------------------------------
// Persistence
// -----------------------------------------------------------------------------

/** Writes the settings through PlayerPrefs. */
export function saveGraphicsSettings(settings, prefs) {
    prefs.setString(PREFS_KEY, JSON.stringify(settings));
    prefs.save();
}

/**
 * Reads the settings back, or null when the player has never chosen any.
 *
 * Null rather than defaults, because "never chosen" is what triggers the capability
 * probe on a first run -- and a probe that ran every launch would silently undo the
 * choice of anybody who had deliberately turned something down.
 */
export function loadGraphicsSettings(prefs) {
    const json = prefs.getString(PREFS_KEY, '');
    if (!json) return null;

    try {
        return Object.assign(new GraphicsSettings(), JSON.parse(json));
    } catch {
        return null;   // A corrupt blob is a first run, not a crash.
    }
}

File: SexyBiscuit.Engine/Rendering/GraphicsSettings.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SexyBiscuit.Engine.Rendering;

// =============================================================================
// Vocabulary
// =============================================================================

/// <summary>How much a shadow costs. The browser engine knows the same list, in order.</summary>
public enum ShadowQuality { Off, Low, Medium, High, Ultra }

/// <summary>How a texture is sampled. Anisotropic falls back where it is unsupported.</summary>
public enum TextureFiltering { Point, Bilinear, Trilinear, Anisotropic }

/// <summary>How much the 2D light map costs, which is mostly its resolution.</summary>
public enum Lighting2DQuality { Off, Low, Medium, High }

// =============================================================================
// The settings
// =============================================================================

/// <summary>
/// Every quality knob the engine has, in one place. The mirror of
/// <c>html5/src/rendering/GraphicsSettings.js</c>.
/// </summary>
/// <remarks>
/// <para>
/// One object rather than a scattering of properties on the renderer, the config and the
/// device manager, because a settings menu has to be able to read them all, write them all
/// and put them in a file. <see cref="Apply"/> is the only place that knows where each one
/// actually lives.
/// </para>
/// <para>
/// The native-only and web-only groups both exist on both engines on purpose. A web export
/// is produced on a desktop, and a preset that lost its <c>MaxPixelRatio</c> on the way
/// through would ship a phone build at three times the resolution it can fill. The
/// capability probe, not this type, decides what a player is shown.
/// </para>
/// </remarks>
public sealed class GraphicsSettings
{
    // -------------------------------------------------------------------------
    // Shared — both engines honour these, or hide them when they cannot
    // -------------------------------------------------------------------------

    /// <summary>The preset these values came from, or "custom" once one is changed.</summary>
    public string Preset { get; set; } = "medium";

    /// <summary>Fraction of the back buffer the world is rendered at, then upscaled.</summary>
    public float RenderScale { get; set; } = 1f;

    public bool VSync { get; set; } = true;

    /// <summary>Frames a second, or 0 for uncapped. Ignored while VSync is on.</summary>
    public int FrameCap { get; set; }

    public ShadowQuality Shadows { get; set; } = ShadowQuality.Low;
    public float ShadowDistance { get; set; } = 50f;

    /// <summary>Lights per draw. The HLSL array is four long and the GLSL one is eight.</summary>
    public int MaxLightsPerObject { get; set; } = 4;

    public int TessellationRadial { get; set; } = 24;
    public int TessellationRings  { get; set; } = 16;

    /// <summary>Far clip, in world units.</summary>
    public float DrawDistance { get; set; } = 1000f;

    public bool FrustumCulling { get; set; } = true;
    public bool SortOpaqueFrontToBack { get; set; } = true;
    public bool Skybox { get; set; } = true;

    /// <summary>Multiplies every emitter's rate, so a preset thins particles rather than killing them.</summary>
    public float ParticleDensity { get; set; } = 1f;

    public bool PostProcessing { get; set; } = true;
    public bool Bloom       { get; set; } = true;
    public bool Vignette    { get; set; }
    public bool ColourGrade { get; set; } = true;
    public bool Scanline    { get; set; }

    public Lighting2DQuality Lighting2D { get; set; } = Lighting2DQuality.Medium;

    public TextureFiltering TextureFiltering { get; set; } = TextureFiltering.Trilinear;
    public int Anisotropy { get; set; } = 4;

    public int StreamingRadius  { get; set; } = 2;
    public int MaxLoadsPerFrame { get; set; } = 1;

    // -------------------------------------------------------------------------
    // Native only
    // -------------------------------------------------------------------------

    /// <summary>Multisample count. 1 is off; the device clamps to what it supports.</summary>
    public int Msaa { get; set; } = 2;

    /// <summary>Square shadow map edge, in texels.</summary>
    public int ShadowMapSize { get; set; } = 1024;

    public bool Fullscreen { get; set; }

    /// <summary>Back-buffer size, or 0 to leave the window alone.</summary>
    public int ResolutionWidth  { get; set; }
    public int ResolutionHeight { get; set; }

    // -------------------------------------------------------------------------
    // Web only
    // -------------------------------------------------------------------------

    /// <summary>Ceiling on devicePixelRatio. Phones report 3 or 4 and cannot fill it.</summary>
    public float MaxPixelRatio { get; set; } = 2f;

    public bool HighDpi { get; set; } = true;

    /// <summary>Context-creation flags. Changing either needs the canvas rebuilt.</summary>
    public bool   Antialias       { get; set; } = true;
    public string PowerPreference { get; set; } = "high-performance";

    /// <summary>Float precision the fragment shaders are compiled at.</summary>
    public string ShaderPrecision { get; set; } = "highp";

    /// <summary>Whether a hidden tab stops rendering, which is most of a laptop's battery.</summary>
    public bool ThrottleWhenHidden { get; set; } = true;

    // -------------------------------------------------------------------------
    // Presets
    // -------------------------------------------------------------------------

    /// <summary>The preset table, shared with the browser engine as one JSON file.</summary>
    public static GraphicsPresetTable Presets => GraphicsPresetTable.Shared;

    /// <summary>Overwrites every field from a preset. Unknown ids are ignored.</summary>
    public bool ApplyPreset(string id)
    {
        GraphicsPreset? preset = Presets.Find(id);
        if (preset is null) return false;

        preset.CopyTo(this);
        Preset = preset.Id;
        return true;
    }

    /// <summary>A copy, for previewing a change without committing to it.</summary>
    public GraphicsSettings Clone() => (GraphicsSettings)MemberwiseClone();

    // -------------------------------------------------------------------------
    // Applying
    // -------------------------------------------------------------------------

    /// <summary>
    /// Pushes every setting to whatever actually owns it.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The one place that knows where each knob lives, which is the reason this type
    /// exists: they are otherwise spread across the two renderers, the mesh cache, the
    /// camera, every emitter and the device manager.
    /// </para>
    /// <para>
    /// Back-buffer settings — resolution, fullscreen, vsync, MSAA — are not applied here.
    /// They need the <c>GraphicsDeviceManager</c>, which belongs to whatever owns the
    /// window, so they go out through <see cref="Core.EngineHost.ApplyDisplaySettings"/>.
    /// Nothing in this repository ever called <c>ApplyChanges</c> before, so changing a
    /// resolution at runtime was not merely unimplemented but impossible.
    /// </para>
    /// </remarks>
    public void Apply(Core.EngineHost? host)
    {
        ApplyGlobals();

        if (host is null) return;

        ApplyTo(host.Renderer3D);
        ApplyTo(host.Renderer2D);
        host.ApplyDisplaySettings?.Invoke(this);
    }

    /// <summary>The settings that live on statics, and so apply with no host at all.</summary>
    public void ApplyGlobals()
    {
        // The primitive cache is keyed on the shape, not on the tessellation, so meshes
        // built at the old segment counts have to be dropped or the change is invisible
        // until something asks for a shape nothing has requested yet.
        if (PrimitiveMesh.RadialSegments != TessellationRadial ||
            PrimitiveMesh.RingSegments   != TessellationRings)
        {
            PrimitiveMesh.RadialSegments = Math.Max(3, TessellationRadial);
            PrimitiveMesh.RingSegments   = Math.Max(2, TessellationRings);
            PrimitiveMesh.ClearCache();
        }

        ParticleSystem3D.DensityScale = ParticleDensity;
        ParticleEmitter.DensityScale  = ParticleDensity;

        foreach (Light3D light in Light3D.All)
        {
            light.CastsShadows  = light.CastsShadows && Shadows != ShadowQuality.Off;
            light.ShadowMapSize = ShadowMapSize;
        }

        if (Camera3D.Main is { } camera) camera.FarClip = DrawDistance;
    }

    private void ApplyTo(RenderSystem3D renderer)
    {
        renderer.EnableShadows        = Shadows != ShadowQuality.Off;
        renderer.ShadowDistance       = ShadowDistance;
        renderer.MaxLightsPerObject   = Math.Clamp(MaxLightsPerObject, 1, 8);
        renderer.EnableFrustumCulling = FrustumCulling;
        renderer.SortOpaqueFrontToBack = SortOpaqueFrontToBack;
        renderer.RenderSkybox         = Skybox;
        renderer.RenderScale          = Math.Clamp(RenderScale, 0.25f, 2f);

        foreach (PostProcessPass3D pass in renderer.PostProcess)
            pass.Enabled = PostProcessing && IsPassEnabled(pass.Name);
    }

    private void ApplyTo(RenderSystem2D renderer)
    {
        renderer.SamplerState = TextureFiltering switch
        {
            TextureFiltering.Point       => Microsoft.Xna.Framework.Graphics.SamplerState.PointClamp,
            TextureFiltering.Anisotropic => Anisotropic(Anisotropy),
            _                            => Microsoft.Xna.Framework.Graphics.SamplerState.LinearClamp,
        };
    }

    /// <summary>Whether a named post-processing pass survives this quality level.</summary>
    private bool IsPassEnabled(string name) => name switch
    {
        "Bloom"       => Bloom,
        "Vignette"    => Vignette,
        "ColourGrade" => ColourGrade,
        "Scanline"    => Scanline,
        _             => true,
    };

    /// <summary>
    /// A cached anisotropic sampler. Cached because a new SamplerState per frame is a
    /// device object per frame, and the renderer reads this every draw.
    /// </summary>
    private static Microsoft.Xna.Framework.Graphics.SamplerState Anisotropic(int level)
    {
        int clamped = Math.Clamp(level, 1, 16);
        if (_anisotropic?.MaxAnisotropy == clamped) return _anisotropic;

        return _anisotropic = new Microsoft.Xna.Framework.Graphics.SamplerState
        {
            Filter        = Microsoft.Xna.Framework.Graphics.TextureFilter.Anisotropic,
            MaxAnisotropy = clamped,
            AddressU      = Microsoft.Xna.Framework.Graphics.TextureAddressMode.Clamp,
            AddressV      = Microsoft.Xna.Framework.Graphics.TextureAddressMode.Clamp,
        };
    }

    private static Microsoft.Xna.Framework.Graphics.SamplerState? _anisotropic;

    // -------------------------------------------------------------------------
    // Persistence
    // -------------------------------------------------------------------------

    /// <summary>The key the whole object is stored under, as one JSON blob.</summary>
    public const string PrefsKey = "Graphics";

    /// <summary>Writes the settings to <see cref="Save.PlayerPrefs"/>.</summary>
    public void Save()
    {
        Engine.Save.PlayerPrefs.SetString(PrefsKey, JsonSerializer.Serialize(this, JsonOptions));
        Engine.Save.PlayerPrefs.Save();
    }

    /// <summary>
    /// Reads the settings back, or returns null when the player has never chosen any.
    /// </summary>
    /// <remarks>
    /// Null rather than defaults, because "never chosen" is what triggers the capability
    /// probe on a first run — and a probe that ran every launch would silently undo the
    /// choice of anybody who had deliberately turned something down.
    /// </remarks>
    public static GraphicsSettings? Load()
    {
        string json = Engine.Save.PlayerPrefs.GetString(PrefsKey, "");
        if (string.IsNullOrWhiteSpace(json)) return null;

        try { return JsonSerializer.Deserialize<GraphicsSettings>(json, JsonOptions); }
        catch (JsonException) { return null; }   // A corrupt blob is a first run, not a crash.
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        Converters = { new JsonStringEnumConverter() },
    };
}
